Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,701 @@
+- `internal function AsNDArray7<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [||||||'TElement||||||]`
+- `internal function AsNDArray8<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [|||||||'TElement|||||||]`
+- `internal function NDArrayData<'TElement, 'TArray>(data : 'TArray) : ['TElement]`
+- `internal function NDArrayStrides<'TArray>(data : 'TArray) : [Int]`
+- `internal function NDArrayShape<'TArray>(data : 'TArray) : [Int]`
+- `internal function NDArrayOffset<'TArray>(data : 'TArray) : Int`
+
+Each of these five functions would be `body intrinsic;`, and together form the contract between the runtime and the core Q# libraries needed to support this proposed feature (see Example 7, below). By necessity, these functions are "unsafe," in that direct access to these functions would allow violating invariants of multidimensional arrays or bypass the type system to expose runtime failures, necessitating the `internal` modifier.
+
+### Examples
+
+Example 1:
+Declaring and indexing into variables of type `[|Double|]` and `[||Int||]` using literals.
+
+```qsharp
+let z = [
+    // Inside [] delimiters, [| |] delimiters refer not to array expressions, but
+    // to "rows" of the two-dimensional array literal.
+    [| 1.0, 0.0 |],
+    [| 0.0, -1.0 |]
+];
+Message($"{z[0, 1]}"); // ← 0.0
+
+// It is a compile-time error for the "rows" denoted by [| ... |] to be uneven:
+let incorrect = [
+    [| 1.0, 2.0 |],
+    [| 10.0, 20.0, 30.0 |] // ← error, since this would declare a "ragged" array
+];
+
+let data = [
+    // We can nest two levels of [| ... |] to get three-dimensional arrays.
+    [|
+        [|0, 1|],
+        [|2, 3|]
+    |],
+
+    [|
+        [|4, 5|],
+        [|6, 7|]
+    |]
+];
+Message($"{data[0, 1, 0]}"); // ← 6
+```
+
+Example 2:
+Mixing 1D and multidimensional arrays.
+
+```qsharp
+// Here, we declare the first two levels of indexing as being rectangular,
+// resulting in a three-dimensional array.
+let data = [
+    [|
+        // After rows, [] denote arrays as elements again.
+        [| [0], [1, 2] |],
+        [| [3, 4, 5], [6, 7, 8, 9] |]
+    |],
+
+    [|
+        [| [10, 11, 12, 13, 14], [15, 16, 17, 18, 19, 20] |],
+        [| [21, 22, 23, 24, 25, 26, 27], [28, 29, 30, 31, 32, 33, 34, 35] |]
+    |]
+];
+// data: [||[Int]||] (that is, a three-dimensional array of arrays of integers)
+// Get the zero-th "plane," first "row", zeroth "column," and third element.
+Message($"{data[0, 1, 0][2]}"); // ← 5
+```
+
+Example 3:
+Using expressions as subarrays of multidimensional arrays results in a compile-time error.
+
+```qsharp
+let a = [2, 3];
+// The following is a compile-time error, since `a` is not a 1D array literal
+// of length 2.
+let data = [
+    [| 0, 1 |],
+    a
+];
+// Using a new library function that can `fail` at runtime works, however.
+let data = Concatenated(0, // concatenate along the 0th (row) axis
```

Comment:
It would be nice to list the public library functions for array construction as well (of course the list won't be binding, and will be reviewed as part of the libraries API meeting).